### DISCUSSION TITLE
spaCy Project: Part-of-speech Tagging & Dependency Parsing

### DISCUSSION THREAD

**Discussion by kanayer** [explosion/spaCy#9916] [Help: Coding & Implementations] [answered]
I am currently working with this [spacy project template](https://github.com/explosion/projects/tree/v3/pipelines/tagger_parser_ud). I was wondering if it is possible to train lemmatizer along with existing POS tagger, morphologizer and dependency parser? Currently, the pipeline in the config file is the following:
`["tok2vec","tagger","morphologizer","parser"]`

The components field in the config file looks like this:

```
[components]

[components.morphologizer]
factory = "morphologizer"

[components.morphologizer.model]
@architectures = "spacy.Tagger.v1"
nO = null

[components.morphologizer.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "*"

[components.parser]
factory = "parser"
learn_tokens = false
min_action_freq = 30
moves = null
update_with_oracle_cut_size = 100

[components.parser.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "parser"
extra_state_tokens = false
hidden_width = 128
maxout_pieces = 3
use_upper = true
nO = null

[components.parser.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "*"

[components.tagger]
factory = "tagger"

[components.tagger.model]
@architectures = "spacy.Tagger.v1"
nO = null

[components.tagger.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "*"

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v1"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v1"
width = ${components.tok2vec.model.encode.width}
attrs = ["LOWER","PREFIX","SUFFIX","SHAPE"]
rows = [5000,2500,2500,2500]
include_static_vectors = false

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v1"
width = 96
depth = 4
window_size = 1
maxout_pieces = 3
```
If it is possible to train a lemmatizer together with the rest of the pipeline, could you please help me understand what kind of component to write into the config file and what kind of changes need to be made?

**Comment by adrianeboyd** [comment, marked as answer]
There is not currently a trainable lemmatizer in the core spacy library. There's an experimental trainable edit tree lemmatizer in development, see:

https://explosion.ai/blog/edit-tree-lemmatizer

There's a UD benchmark project that uses it (also with a trainable tokenizer that I wouldn't recommend outside of benchmarking, see https://explosion.ai/blog/ud-benchmarks-v3-2):

https://github.com/explosion/projects/tree/v3/benchmarks/ud_benchmark

I would currently guess that the edit tree lemmatizer could move into the core library in v3.3.0, but we haven't made an official decision yet.

**Reply by kanayer** [reply]
How did the English model get the lemmatizer then? It is possible to see the lemmas with `token.lemma_` for English model.

**Reply by adrianeboyd** [reply]
The current default English lemmatizer is a rule-based lemmatizer with hand-written rules. There is support for lookup, POS-based lookup, or rule-based lemmatizers for a number of languages from data in the package `spacy-lookups-data`. And a few languages use external libraries for lemmas (Russian and Ukrainian in a separate lemmatizer component, Japanese and Korean as part of the tokenizer).

If you add this, you should get the default lemmatizer (which depends on the language) if one is available, or otherwise errors that the relevant lookup tables are missing:

```ini
[components.lemmatizer]
factory = "lemmatizer"
```

You can also use:

```python
nlp.add_pipe("lemmatizer").initialize()
```

To try out the same thing, or add extra config settings to try other modes (like `lookup` if the default is `rule`). See an example in the section "Switch from rule-based to lookup lemmatization" here: https://spacy.io/models#design-modify

**Reply by vistamou** [reply]
Dear @adrianeboyd,

I have a question regarding the experimental lemmatizer; I tried to run it following the instructions
provided here: https://explosion.ai/blog/edit-tree-lemmatizer.
But I get the following error:

`ValueError: [E002] Can't find factory for 'experimental_edit_tree_lemmatizer' for language Bulgarian (bg). This usually happens when spaCy calls `nlp.create_pipe` with a custom component name that's not registered on the current language class. If you're using a Transformer, make sure to install 'spacy-transformers'. If you're using a custom component, make sure you've added the decorator `@Language.component` (for function components) or `@Language.factory` (for class components).`

It seems that it doesn't recognizer the language, but I can't understand why.

I'm attaching also part of the config file: 

```
[nlp]
lang = "bg"
pipeline = ["tok2vec", "senter", "tagger","morphologizer", "experimental_edit_tree_lemmatizer"]
batch_size = 1000
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.experimental_edit_tree_lemmatizer]
factory = "experimental_edit_tree_lemmatizer"

[components.experimental_edit_tree_lemmatizer.model]
@architectures = "spacy.Tagger.v1"
nO = null

[components.experimental_edit_tree_lemmatizer.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "tok2vec"


```

Could you please help? Thank you!!

**Reply by adrianeboyd** [reply]
Did you follow the install steps here? https://explosion.ai/blog/edit-tree-lemmatizer#try-out

**Reply by vistamou** [reply]
yes exaclty, I skipped **only** the sample project part--
but when I'm running the command `pip list` I don't see
any version of 'spacy-experimental'.

I suppose to see something like that?

**Reply by kanayer** [reply]
> The current default English lemmatizer is a rule-based lemmatizer with hand-written rules. There is support for lookup, POS-based lookup, or rule-based lemmatizers for a number of languages from data in the package `spacy-lookups-data`. And a few languages use external libraries for lemmas (Russian and Ukrainian in a separate lemmatizer component, Japanese and Korean as part of the tokenizer).
> 
> If you add this, you should get the default lemmatizer (which depends on the language) if one is available, or otherwise errors that the relevant lookup tables are missing:
> 
> ```ini
> [components.lemmatizer]
> factory = "lemmatizer"
> ```
> 
> You can also use:
> 
> ```python
> nlp.add_pipe("lemmatizer").initialize()
> ```
> 
> To try out the same thing, or add extra config settings to try other modes (like `lookup` if the default is `rule`). See an example in the section "Switch from rule-based to lookup lemmatization" here: https://spacy.io/models#design-modify

If I am working on the Korean model, how can use the lemmatizer incorporated into the Korean tokenizer? Also, how can I add the Korean tokenizer into the config file?

**Reply by adrianeboyd** [reply]
You have to have `spacy-experimental` installed for the experimental lemmatizer to be found. If it's not in `pip list`, something has gone wrong, try installing it again.

The segmentation across different Korean resources is an issue, as we've discussed in other threads. You can use the standard rule-based tokenizer for the UD Korean corpora by specifying it (`spacy.Tokenizer.v1`) directly in the config by hand. (You can't use `spacy init config -l ko` to get this configuration. It requires manual editing.)

These UD benchmark examples aren't using the same tokenizer (`spacy.Tokenizer.v1` would have worked fine for Korean, but I wanted to use the same config with a trainable tokenizer for all languages), but you can see the edit tree lemmatizer in action for Korean here:

https://huggingface.co/explosion/ko_udv25_koreankaist_trf

There's no good way to combine lemmas from mecab from `spacy.ko.KoreanTokenizer` with a pipeline trained on the (current) UD Korean corpora because the tokenization is very very different.

**Reply by kanayer** [reply]
Wow, the benchmark model looks really cool! If I have my own dataset with lemmas, pos tags, head info, and dependency tags in `conllu ` format (bigger than KAIST or GSD) can I re-train the model? Can I put my dataset into the assets folder in the ud_bechmark project that i cloned using `python -m spacy project clone benchmarks/ud_benchmark`

**Reply by adrianeboyd** [reply]
As long as the paths/names work out for the steps in `project.yml`, it should work fine.

Because of the current defaults for `ko`, you have to have mecab installed even though it isn't used in the end. I hope to improve this in v3.3, but it's still on my to-do list.

**Reply by kanayer** [reply]
Wow, this is really awesome! I'm fascinated by the spaCy NLP developers team! 
Do you think this will work on the Thai dataset? 
How much GPU memory did you need to train these models with big datasets?
Also, is it normal that displaCy throws an error when I use it with the text analyzed by `ko_udv25_koreankaist_trf`?

**Reply by adrianeboyd** [reply]
Typically ~12-15GB, but with the default settings it depends on the size of the dev set mainly. You can always adjust the batch sizes if necessary, though.

The only UD Thai corpus I see is way too small.

displacy works fine on my end?

**Reply by adrianeboyd** [reply]
The built-in huggingface demo doesn't show anything, though, because it only supports NER.
